tools/proteore_reactome/reactome_analysis.py

```diff
@@ -0,0 +1,116 @@
+import argparse
+import csv
+import json
+import os
+import re
+
+CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+def id_valid(identifiers):
+    """
+    Validate IDs if they contain special characters
+    """
+    res = []
+    remove = []
+    for id in identifiers:
+        id = id.split(";")[0]
+        if re.match("^[A-Za-z0-9_-]*$", id):
+            res.append(id)
+        else:
+            remove.append(id)
+    return res, remove
+
+
+def isnumber(format, n):
+    """
+    Check if an variable is numeric
+    """
+    float_format = re.compile(r"^[-]?[1-9][0-9]*.?[0-9]+$")
+    int_format = re.compile(r"^[-]?[1-9][0-9]*$")
+    test = ""
+    if format == "int":
+        test = re.match(int_format, n)
+    elif format == "float":
+        test = re.match(float_format, n)
+    if test:
+        return True
+    else:
+        return False
+
+
+def data_json(identifiers):
+    """
+    Submit IDs list to Reactome and return results in json format
+    Return error in HTML format if web service is not available
+    """
+    trash = []
+    if identifiers[1] == "list":
+        ids = identifiers[0].split()
+        ids = [x.split(";") for x in ids]
+        ids = [item.strip() for sublist in ids for item in sublist if item != '']  # noqa 501
+        json_string = os.popen("curl -H \"Content-Type: text/plain\" -d \"$(printf '%s')\" -X POST --url www.reactome.org/AnalysisService/identifiers/\?pageSize\=1\&page\=1" % ids).read()  # noqa 501
+        if len(id_valid(identifiers[0].split())[1]) > 0:
+            trash = id_valid(identifiers[0].split())[1]
+    elif identifiers[1] == "file":
+        header = identifiers[2]
+        with open(identifiers[0], "r") as mq:
+            file_content = csv.reader(mq, delimiter="\t")
+            file_content = list(file_content)   # csv object to list
+            ncol = identifiers[3]
+            if isnumber("int", ncol.replace("c", "")):
+                if header == "true":
+                    idens = [x for x in [line[int(ncol.replace("c", ""))-1].split(";") for line in file_content[1:]]]  # noqa 501
+                else:
+                    idens = [x for x in [line[int(ncol.replace("c", ""))-1].split(";") for line in file_content]]  # noqa 501
+                # flat list of list of lists, remove empty items
+                idens = [item.strip() for sublist in idens for item in sublist if item != '']  # noqa 501
+                ids = "\n".join(id_valid(idens)[0])
+                json_string = os.popen("curl -H \"Content-Type: text/plain\" -d \"$(printf '%s')\" -X POST --url www.reactome.org/AnalysisService/identifiers/\?pageSize\=1\&page\=1 2> stderr" % ids).read()  # noqa 501
+                if len(id_valid(idens)[1]) > 0:
+                    trash = id_valid(idens)[1]
+    # print(json_string)
+    j = json.loads(json_string)
+    print("Identifiers not found: " + str(j["identifiersNotFound"]))
```

Comment:
you could use f-strings, they are super cool: https://tirkarthi.github.io/programming/2019/05/08/f-string-debugging.html